lib/SILOptimizer/Analysis/AccessedStorageAnalysis.cpp

```diff
@@ -94,42 +94,57 @@ bool FunctionAccessedStorage::updateUnidentifiedAccess(
   return updateAccessKind(unidentifiedAccess.getValue(), accessKind);
 }
 
+// Merge the given FunctionAccessedStorage in `other` into this
+// FunctionAccessedStorage. Use the given `transformStorage` to map `other`
+// AccessedStorage into this context. If `other` is from a callee, argument
+// substitution will be performed if possible. However, there's no guarantee
+// that the merged access values will belong to this function.
+//
+// Note that we may have `this` == `other` for self-recursion. We still need to
+// propagate and merge in that case in care arguments are recursively dependent.
 bool FunctionAccessedStorage::mergeAccesses(
-    const FunctionAccessedStorage &RHS,
+    const FunctionAccessedStorage &other,
     std::function<AccessedStorage(const AccessedStorage &)> transformStorage) {
 
   bool changed = false;
-
-  for (auto &accessEntry : RHS.storageAccessMap) {
-
-    const AccessedStorage &storage = transformStorage(accessEntry.first);
-    // transformStorage returns invalid storage object for local storage
-    // that should not be merged with the caller.
-    if (!storage)
-      continue;
-
-    if (storage.getKind() == AccessedStorage::Unidentified)
-      changed |= updateUnidentifiedAccess(accessEntry.second.accessKind);
-    else {
-      auto result = storageAccessMap.try_emplace(storage, accessEntry.second);
-      if (result.second)
-        changed = true;
-      else
-        changed |= result.first->second.mergeFrom(accessEntry.second);
+  bool restart;
+  do {
+    restart = false;
+    for (auto &accessEntry : other.storageAccessMap) {
+
+      const AccessedStorage &storage = transformStorage(accessEntry.first);
+      // transformStorage returns invalid storage object for local storage
+      // that should not be merged with the caller.
+      if (!storage)
+        continue;
+
+      if (storage.getKind() == AccessedStorage::Unidentified)
+        changed |= updateUnidentifiedAccess(accessEntry.second.accessKind);
+      else {
+        auto result = storageAccessMap.try_emplace(storage, accessEntry.second);
+        if (result.second) {
+          changed = true;
+          // Insertion in DenseMap invalidates the iterator (if `this` ==
+          // `other`. Break out of the loop and start over.
+          restart = true;
```

Comment:
Did you test the impact of this change on compile time? I am concerned about the worst-case complexity of this change / data structure: we might have a lot of repeated / unnecessary work if we frequently restart this loop.

Reply:
The set size should be tiny here. I just couldn't stand copying data to handle an exceedingly rare corner case. But, there's no sense having confusing code. Changed to...

```
  // Insertion in DenseMap invalidates the iterator in the rare case of
  // self-recursion (`this` == `other`) that passes accessed storage though an
  // argument. Rather than complicate the code, make a temporary copy of the
  // AccessedStorage.
  SmallVector<std::pair<AccessedStorage, StorageAccessInfo>, 8> otherAccesses;
  otherAccesses.reserve(other.storageAccessMap.size());
  otherAccesses.append(other.storageAccessMap.begin(),
                       other.storageAccessMap.end());

  bool changed = false;
  for (auto &accessEntry : otherAccesses) {
```

Reply:
Regarding compile time. This pass should not make a measurable difference. I'm sanity checking with some real-world projects.

My general philosophy, which drove the decisions in this pass and the upcoming whole module pass:
- Minimize the memory consumption and cost of the no-optimization case. Additional costs are proportional to the amount of optimization actually performed. Typically this means a single pass over the IR (possibly relying on DFS) doing an opcode check and a single hash table of optimizable things. This pass relies on interprocedural analysis before doing any optimization, so I was paranoid about its cost and size of the results.
- Worst-case, pathological behavior, which we never see in testing, is more important to guard against than it is to reduce typical overhead of the optimization.